File: 백준/Gold/16935. 배열 돌리기 3/배열 돌리기 3.py
import sys
input = sys.stdin.readline

height,width,R = map(int,input().split(' '))

A = [list(map(int,input().split(' '))) for _ in range(height)]

operation = list(map(int,input().split(' ')))

def measure(arr):
  height = len(arr)
  width = len(arr[0])
  height_half = height//2
  width_half = width//2
  return (height,width,height_half,width_half)

def op_1(arr):
  return arr[::-1]

def op_2(arr):
  temp = []
  for row in arr:
      temp.append(row[::-1])
  return temp

def op_3(arr):
  height,width,_,_ = measure(arr)
  temp = [[0 for _ in range(height)] for _ in range(width)]
  for i in range(0,width):
    for j in range(0,height):
      temp[i][j] = arr[height-(j+1)][i]
  return temp

def op_4(arr):
  height,width,_,_ = measure(arr)
  temp = [[0 for _ in range(height)] for _ in range(width)]
  for i in range(0,width):
    for j in range(0,height):
      temp[i][j] = arr[j][width-(i+1)]
  return temp

def op_5(arr):
  height,width,height_half,width_half = measure(arr)
  temp = [[0 for _ in range(width)]for _ in range(height)]
  for i in range(0,height_half):
    for j in range(0,width_half):
      temp[i][j] = arr[i+height_half][j]       
      temp[i][j+width_half] = arr[i][j]       
      temp[i+height_half][j+width_half] = arr[i][j+width_half]       
      temp[i+height_half][j] = arr[i+height_half][j+width_half]
  return temp

def op_6(arr):
  height,width,height_half,width_half = measure(arr)
  temp = [[0 for _ in range(width)]for _ in range(height)]
  for i in range(0,height_half):
    for j in range(0,width_half):
      temp[i][j] = arr[i][j+width_half]       
      temp[i][j+width_half] = arr[i+height_half][j+width_half]       
      temp[i+height_half][j+width_half] = arr[i+height_half][j]       
      temp[i+height_half][j] = arr[i][j]
  return temp

for op in operation:
  if op == 1:
    A = op_1(A)
  elif op == 2:
    A = op_2(A)
  elif op == 3:
    A = op_3(A)
  elif op == 4:
    A = op_4(A)
  elif op == 5:
    A = op_5(A)
  elif op == 6:
    A = op_6(A)

for row in A:
  print(' '.join(map(str,row)))

File: 백준/Gold/16935. 배열 돌리기 3/배열 돌리기 3.js
// 입력
const fs = require("fs");
const inputPath = process.platform === "linux" ? "/dev/stdin" : "./test.txt";
const input = fs.readFileSync(inputPath).toString().trim().split('\n');

const [N, M, R] = input[0].split(' ').map(Number);
let arr = [];
for (let i = 1 ; i <= N ; i++) {
    arr.push(input[i].trim().split(' '));
}
const cal = input[N+1].split(' ');

// 로직
function reverseUpdown() {
    const height = arr.length;
    for (let i = 0 ; i < Math.floor(height/2) ; i++) {
        const temp = arr[i];
        arr[i] = arr[height-i-1];
        arr[height-i-1] = temp;
    }
}

function reverseLeftRight() {
    const height = arr.length;
    for (let i = 0 ; i < height ; i++) {
        arr[i].reverse();
    }
}

function rotateRight() {
    const height = arr.length;
    const width = arr[0].length;
    
    const newArr = Array.from({length: width}, () => Array(height).fill(0));
    for (let i = 0 ; i < height ; i++) {
        for (let j = 0 ; j < width ; j++) {
            newArr[j][height-i-1] = arr[i][j];
        }
    }

    arr = newArr;
}

function rotateLeft() {
    const height = arr.length;
    const width = arr[0].length;
    
    const newArr = Array.from({length: width}, () => Array(height).fill(0));
    for (let i = 0 ; i < height ; i++) {
        for (let j = 0 ; j < width ; j++) {
            newArr[width-j-1][i] = arr[i][j];
        }
    }

    arr = newArr;
}

function moveClockwise() {
    const height = arr.length;
    const width = arr[0].length;

    const hh = Math.floor(height/2);
    const hw = Math.floor(width/2);
    const newArr = Array.from({length: height}, () => Array(width).fill(0));
    
    for (let i = 0 ; i < hh ; i++) {
        for (let j = 0 ; j < hw ; j++) {
            newArr[i][j+hw] = arr[i][j];
            newArr[i][j] = arr[i+hh][j];
            newArr[i+hh][j] = arr[i+hh][j+hw];
            newArr[i+hh][j+hw] = arr[i][j+hw];
        }
    }

    arr = newArr;
}

function moveCounterClockwise() {
    const height = arr.length;
    const width = arr[0].length;

    const hh = Math.floor(height/2);
    const hw = Math.floor(width/2);
    const newArr = Array.from({length: height}, () => Array(width).fill(0));
    for (let i = 0 ; i < hh ; i++) {
        for (let j = 0 ; j < hw ; j++) {
            newArr[i][j] = arr[i][j+hw];
            newArr[i+hh][j] = arr[i][j];
            newArr[i+hh][j+hw] = arr[i+hh][j];
            newArr[i][j+hw] = arr[i+hh][j+hw];
        }
    }

    arr = newArr;
}

for (let c of cal) {
    if (c === '1') {
        reverseUpdown();
    }
    else if (c === '2') {
        reverseLeftRight();
    }
    else if (c === '3') {
        rotateRight();
    }
    else if (c === '4') {
        rotateLeft();
    }
    else if (c === '5') {
        moveClockwise();
    }
    else {
        moveCounterClockwise();
    }

}
const answer = arr.map((row)=>row.join(' ')).join('\n');
console.log(answer);